Add GET param support

LucasRoesler opened the pull request:
**What**
- Search the `token` GET parameter for a valid JWT
- Return the token prefix
- Make the test suit part of the `jwt` package so that the tests are
easier to run and get coverage information from `go test`
- Implement `GetUnvalidatedClaims` method

**Why**
- It is normal to support GET parameters for tokens
- The token prefix can be important information about how the token
should be treated, e.g. `Bearer` tokens may require one set of claims but
a `Resource` token may require a different set of claims.
- `GetUnvalidatedClaims` allows the caller to try and extract the claims while ignoring
signiture validation errors.  Specifically, it allows someone to extract
the claims without needing the signing key

LucasRoesler committed:
Add GET param support

**What**
- Search the `token` GET parameter for a valid JWT
- Return the token prefix
- Make the test suit part of the `jwt` package so that the tests are
easier to run and get coverage information from `go test`

**Why**
- It is normal to support GET parameters for tokens
- The token prefix can be important information about how the token
should be treated, e.g. `Bearer` tokens may require one set of claims but
a `Resource` token may require a different set of claims.

Signed-off-by: Lucas Roesler <roesler.lucas@gmail.com>

LucasRoesler commented:
I fixed the coverage change, now it has gone up +0.2%

LucasRoesler committed:
Add method to get claims while ignoreing validation

**What**
- Implement `GetUnvalidatedClaims` method

**Why**
- This allows the caller to try and extract the claims while ignoring
signiture validation errors.  Specifically, it allows someone to extract
the claims without needing the signing key

Signed-off-by: Lucas Roesler <roesler.lucas@gmail.com>

trusch commented on README.md:
````diff
@@ -28,7 +28,7 @@ CreateToken takes some claims and a key (either private rsa, private ec or hmac
 
 ```go
-func GetTokenFromRequest(r *http.Request) (string, error)
+func GetTokenFromRequest(r *http.Request) (TokenPrefix, string, error)
````
TokenPrefix is not a type

trusch commented on README.md:
````diff
@@ -64,7 +64,7 @@ ParsePublicKey parses a pem encoded public key (rsa or ecdsa based)
 
 ```go
-func GetClaimsFromRequest(r *http.Request, key interface{}) (Claims, error)
+func GetClaimsFromRequest(r *http.Request, key interface{}) (TokenPrefix, Claims, error)
````
TokenPrefix is not a type

trusch commented on jwt.go:
```diff
@@ -81,6 +89,27 @@ func ValidateToken(tokenString string, key interface{}) (Claims, error) {
+		return nil, err
+	}
+
+	err = json.NewDecoder(bytes.NewBuffer(claimBytes)).Decode(&claims)
```
bad style, please use one line for the if and the Decode

trusch commented on jwt.go:
```diff
@@ -81,6 +89,27 @@ func ValidateToken(tokenString string, key interface{}) (Claims, error) {
+		return nil, err
+	}
+
+	return claims, nil
```
you could even "return claims, json.NewDecoder(buf).Decode(&claims"

trusch commented on jwt.go:
```diff
@@ -125,21 +154,48 @@ func ParsePrivateKey(data []byte) (interface{}, error) {
 // GetClaimsFromRequest extracts and validates the token from a request, returning the claims
-func GetClaimsFromRequest(r *http.Request, key interface{}) (Claims, error) {
-	token, err := GetTokenFromRequest(r)
+func GetClaimsFromRequest(r *http.Request, key interface{}, skipValidate bool) (prefix string, claims Claims, err error) {
```
perhaps it would be better readable to use a variadic skipValidate parameter so that it can be left out in the default case

LucasRoesler replied:
Seems like an abuse of variadic options

trusch replied:
yes, a bit. But now there would be in 99% of all cases a trailing "false" in the parameter list. Just a question of style, this will not prevent us from merging this PR

LucasRoesler replied:
true, i figured that was just the price we pay for using Go 😄

chris-baynes replied:
Personally I always try to avoid function signatures with bool flag arguments. In most cases I think it's clearer from an api perspective to just have a separate function, e.g. `GetClaimsFromRequestWithValidation`

LucasRoesler replied:
I just pushed a version of the code that looks like this, for review.  If we don't like it we can revert it :)

LucasRoesler committed:
Simplify GetUnvalidatedClaims return

**What**
- Fix typo in the README
- Simplify the return statement in GetUnvalidatedClaims

Signed-off-by: Lucas Roesler <roesler.lucas@gmail.com>

LucasRoesler committed:
Split GetClaimFromRequest into two methods

**What**
- Add an explict method `GetClaimsFromRequestWithValidation`
- Remove the `skipValidation` boolean parameter from
`GetClaimsFromRequest`

**Why**
- This is more explicit and does not include any optional parameters

Signed-off-by: Lucas Roesler <roesler.lucas@gmail.com>
